File: packages/server/ExchangeGame/Game.cs
using ExchangeGame.Messaging;
using ExchangeGame.Messaging.Messages;
using ExchangeGame.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using WatsonWebsocket;

namespace ExchangeGame
{
    public class Game
    {
        public Dictionary<int, Player> Players { get; } = new Dictionary<int, Player>();

        public Dictionary<int, Exchange> Exchanges { get; } = new Dictionary<int, Exchange>();

        public Dictionary<string, Player> PlayersByIp { get; } = new Dictionary<string, Player>();

        public Dictionary<int, Attendee> Attendees { get; set; } = new Dictionary<int, Attendee>();

        public int Score { get; private set; } = 0;

        public WatsonWsServer server { get; set; }

        private Random _random = new Random();

        private Dictionary<Player, HashSet<Attendee>> _availableAttendees = new Dictionary<Player, HashSet<Attendee>>();

        // TODO make this configurable
        private const int FAIL_SCORE = 100;

        // TODO make this configurable
        private const int EXCHANGE_COUNT = 3;

        public Game()
        {

            var exchangeNames = ContentRepository.GetExchangeNames(EXCHANGE_COUNT);
            foreach(var exchangeName in exchangeNames)
            {
                var newExchange = new Exchange(exchangeName, OnUpdateScore);
                Exchanges.Add(newExchange.Id, newExchange);
            }
        }

        public Player AddPlayer(string name = "Test player", string ipPort = "")
        {
            var newPlayer = new Player(name);
            Players.Add(newPlayer.Id, newPlayer);
            PlayersByIp.Add(ipPort, newPlayer);
            newPlayer.SendMessage = messageStr => server.SendAsync(ipPort, messageStr);

            _availableAttendees[newPlayer] = newPlayer.Attendees.ToHashSet();
            foreach (var attendee in newPlayer.Attendees)
            {
                Attendees.Add(attendee.Id, attendee);
            }

            var message = new LobbyMessage(Players.Values);
            BroadcastMessage(JsonHelpers.SerializeMessage(message));
            
            return newPlayer;
        }

        public void PlayerReady(Player player)
        {
            player.Ready = true;
            if (Players.Values.Count > 1 && Players.Values.All(x => x.Ready))
            {
                Start();
                SendinitialCalls();
            }
        }

        public void Connect(int exchangeId, int attendeeId)
        {
            var exchange = Exchanges[exchangeId];
            var attendee = Attendees[attendeeId];
            exchange.Connect(attendee);
        }

        private void Start()
        {
            foreach (var player in Players.Values)
            {
                var message = new StartMessage(Exchanges.Values, player.Attendees);
                player.SendMessage(JsonHelpers.SerializeMessage(message));
            }
        }

        private void SendinitialCalls()
        {
            var messages = new List<string>();
            foreach(var player in Players.Values)
            {
                CreateCall(player);
            }
        }

        private void CreateCall(Player player)
        {
            var call = MatchCall(player);
            var exchange = GetRandomExchange();
            exchange.AddCall(call);
            player.AddCall(call);
            call.OnComplete += OnCallDisposed;
            call.OnTimeout += OnCallDisposed;
            var message = new CallMessage(call, exchange);
            player.SendMessage(JsonHelpers.SerializeMessage(message));
            Console.WriteLine($"Creating new call for ${call.Sender.Player.DisplayName}{call.Sender.Player.Id}");
        }

        private void BroadcastMessage(string message)
        {
            foreach(var player in Players.Values)
            {
                player.SendMessage(message);
            }
        }


        private Exchange GetRandomExchange()
        {
            var exchanges = Exchanges.Values.ToList();
            var idx = _random.Next(0, exchanges.Count());
            return exchanges[idx];
        }

        private void OnUpdateScore(int scoreToUpdate, Exchange exchange, Attendee attendee)
        {
            Score += scoreToUpdate;
            var message = new ScoreMessage(exchange, attendee, Score);
            BroadcastMessage(JsonHelpers.SerializeMessage(message));
        }

        private Call MatchCall(Player player = null)
        {
            var selectedPlayers = Players.Values.OrderBy(x => _random.Next()).Take(2).ToList();
            var senderPlayer = player ?? selectedPlayers[0];
            var recipientPlayer = selectedPlayers[1];

            var sender = _availableAttendees[senderPlayer].ElementAt(_random.Next(0, _availableAttendees[senderPlayer].Count));
            var recipient = _availableAttendees[recipientPlayer].ElementAt(_random.Next(0, _availableAttendees[recipientPlayer].Count));

            _availableAttendees[senderPlayer].Remove(sender);
            _availableAttendees[recipientPlayer].Remove(recipient);

            return new Call(ContentRepository.GetCallNames(1).First(), sender, recipient);
        }

        private void OnCallDisposed(Call call)
        {
            _availableAttendees[call.Sender.Player].Add(call.Sender);
            _availableAttendees[call.Recipient.Player].Add(call.Recipient);

            CreateCall(call.Sender.Player);
        }
    }
}

File: packages/server-v2/src/Game.ts
import { WebSocket } from "ws";
import { Response } from "./api";
import Attendee from "./GameObjects/Attendee";
import Call from "./GameObjects/Call";
import Exchange from "./GameObjects/Exchange";
import Player from "./GameObjects/Player";
import { getCallNames, getExchangeNames, getRandomInt } from "./utils";

export default class Game {
  public gameOver: boolean = false;
  public players: Record<number, Player> = {};
  public playersByConnection: Record<string, Player> = {};
  public connections: Record<number, WebSocket> = {};
  public exchanges: Record<number, Exchange> = {};
  public attendees: Record<number, Attendee> = {};
  public score: number = 0;
  private _availableAttendees: Record<number, Set<Attendee>> = {};
  private FAIL_SCORE = 100;
  private EXCHANGE_COUNT = 3;

  constructor() {
    const exchangeNames = getExchangeNames(this.EXCHANGE_COUNT);
    exchangeNames.forEach((name) => {
      const newExchange = new Exchange(name, this.onUpdateScore);
      this.exchanges[newExchange.id] = newExchange;
      return newExchange;
    });
  }

  public addPlayer(name: string, connectionId: string, ws: WebSocket) {
    const newPlayer = new Player(name, (message: object) =>
      ws.send(JSON.stringify(message))
    );
    this.players[newPlayer.id] = newPlayer;
    this.playersByConnection[connectionId] = newPlayer;

    this._availableAttendees[newPlayer.id] = new Set(newPlayer.attendees);
    newPlayer.attendees.forEach((attendee) => {
      this.attendees[attendee.id] = attendee;
    });

    const message: Response = {
      type: "Lobby",
      payload: {
        players: Object.values(this.players).map((player) => ({
          displayName: player.displayName,
          id: player.id,
          ready: player.ready,
        })),
      },
    };
    this.broadcastMessage(message);
  }

  public playerReady(connectionId: string) {
    const player = this.playersByConnection[connectionId];
    player.ready = true;
    const message: Response = {
      type: "Lobby",
      payload: {
        players: Object.values(this.players).map((player) => ({
          displayName: player.displayName,
          id: player.id,
          ready: player.ready,
        })),
      },
    };
    this.broadcastMessage(message);
    const playerValues = Object.values(this.players);
    if (playerValues.length > 1 && playerValues.every((x) => x.ready)) {
      this.start();
      this.sendInitialCalls();
    }
  }

  public connect(exchangeId: number, attendeeId: number) {
    const exchange = this.exchanges[exchangeId];
    const attendee = this.attendees[attendeeId];
    console.log(
      `connecting ${attendee.displayName} with ${exchange.displayName}`
    );
    exchange.connect(attendee);
  }

  public start() {
    Object.values(this.players).forEach((player) => {
      const message: Response = {
        type: "Start",
        payload: {
          exchanges: Object.values(this.exchanges).map((exchange) => ({
            id: exchange.id,
            displayName: exchange.displayName,
          })),
          attendees: Object.values(player.attendees).map((attendee) => ({
            id: attendee.id,
            displayName: attendee.displayName,
          })),
          score: this.score,
          timestamp: Date.now(),
        },
      };
      player.sendMessage(message);
    });
  }

  private sendInitialCalls() {
    const messages = [];

    Object.values(this.players).forEach((player) => {
      this.createCall(player);
    });
  }

  private createCall(player: Player) {
    const call: Call = this.matchCall(player);
    const exchange = this.getRandomExchange();
    player.addCall(call);
    exchange.addCall(call);
    call.onComplete.push(this.onCallDisposed);
    call.onTimeout.push(this.onCallDisposed);

    const message: Response = {
      type: "Mission",
      payload: {
        callee: {
          id: call.recipient.id,
          displayName: call.recipient.displayName,
        },
        caller: {
          id: call.sender.id,
          displayName: call.sender.displayName,
        },
        displayName: call.displayName,
        duration: call.duration,
        exchange: exchange.id,
        id: call.id,
        timestamp: call.startTimestamp,
      },
    };
    player.sendMessage(message);
    console.log(
      `Creating new call for ${
        this.players[call.sender.playerId].displayName
      } ${this.players[call.sender.playerId].id}`
    );
  }

  private broadcastMessage(message: object) {
    Object.values(this.players).forEach((player) => {
      player.sendMessage(message);
    });
  }

  private getRandomExchange(): Exchange {
    const exchanges = Object.values(this.exchanges);
    return exchanges[getRandomInt(0, exchanges.length)];
  }

  private onUpdateScore = (
    scoreToUpdate: number,
    exchange?: Exchange,
    attendee?: Attendee
  ) => {
    this.score += scoreToUpdate;
    if (this.score >= this.FAIL_SCORE) {
      this.setGameOver();
    } else {
      const message: Response = {
        type: "Score",
        payload: {
          score: this.score,
          attendee: attendee?.id ?? 1,
          exchange: exchange?.id ?? 1,
        },
      };
      this.broadcastMessage(message);
    }
  };

  private setGameOver() {
    this.gameOver = true;
    const message: Response = {
      type: "GameOver",
      payload: { timestamp: Date.now() },
    };
    this.broadcastMessage(message);
  }

  private matchCall(senderPlayer: Player): Call {
    const playersToSelect = Object.values(this.players)
      .map((player) => (player === senderPlayer ? null : player))
      .filter(Boolean) as Player[];

    const recipientPlayer =
      playersToSelect[getRandomInt(0, playersToSelect.length)];

    const sender = Array.from(
      this._availableAttendees[senderPlayer.id].values()
    )[getRandomInt(0, this._availableAttendees[senderPlayer.id].size)];
    const recipient = Array.from(
      this._availableAttendees[recipientPlayer.id].values()
    )[getRandomInt(0, this._availableAttendees[recipientPlayer.id].size)];

    this._availableAttendees[senderPlayer.id].delete(sender);
    this._availableAttendees[senderPlayer.id].delete(recipient);

    return new Call(getCallNames(1)[0], sender, recipient);
  }

  private onCallDisposed = (call: Call) => {
    this._availableAttendees[call.sender.playerId].add(call.sender);
    this._availableAttendees[call.recipient.playerId].add(call.recipient);

    if (!this.gameOver) {
      this.createCall(this.players[call.sender.playerId]);
    }
  };
}
